### PR TITLE
Add s3bucket resource and implement get current and desired states

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#2023

Adds the s3bucket resource and implements getting the current and desired states. I'll implement create and delete as later PRs.

Also extracts getting the IAM user to an `aws` service for AWS logic that needs to be shared between resources. If this looks OK then I'll update the `cloudformation` resource to use it.

### CODE REVIEW

File: service/resource/s3bucketv1/mocks.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	// (splitting by colon)
+	accountIDIndex  = 4
+	accountIDLength = 12
+)
+
+type IAMClientMock struct {
+	accountID string
+}
+
+func (i *IAMClientMock) GetUser(input *iam.GetUserInput) (*iam.GetUserOutput, error) {
```
Comment (rossf7) [inline service/resource/s3bucketv1/mocks.go]: Currently GetUser is duplicated. I think we could move this to an `aws` service. WDYT?

File: service/resource/s3bucketv1/resource.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		return "", microerror.Mask(err)
+	}
+	userArn := *resp.User.Arn
+	accountID := strings.Split(userArn, ":")[accountIDIndex]
+	if err := adapter.ValidateAccountID(accountID); err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return accountID, nil
+}
```
Comment (rossf7) [inline service/resource/s3bucketv1/resource.go]: Again currently getAccountID is duplicated.

File: service/aws/mocks.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package aws
```
Comment (rossf7) [inline service/aws/mocks.go]: Here is the mock for the AWS service.

File: service/aws/service.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
+package aws
```
Comment (rossf7) [inline service/aws/service.go]: The aws service has methods that need to be shared between resources.

File: service/resource/s3bucketv1/resource.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+// Config represents the configuration used to create a new s3bucket resource.
+type Config struct {
+	// Dependencies.
+	AwsService *awsservice.Service
+	Clients    Clients
+	Logger     micrologger.Logger
+
+	// Settings.
+	AwsConfig awsutil.Config
```
Comment (fgimenez) [inline service/resource/s3bucketv1/resource.go]: Do we need the AwsConfig here?

File: service/resource/s3bucketv1/current.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	bucketName := key.BucketName(customObject, accountID)
+	headInput := &s3.HeadBucketInput{
+		Bucket: aws.String(bucketName),
+	}
+	_, err = r.clients.S3.HeadBucket(headInput)
+	if IsBucketNotFound(err) {
+		r.logger.LogCtx(ctx, "debug", "did not find the S3 bucket")
+		return BucketState{}, nil
+
+	} else if err != nil {
```
[the remark this answers is not in the document]
Comment (rossf7) [reply service/resource/s3bucketv1/current.go]: Good point, removed.

File: service/resource/s3bucketv1/resource.go (hunk 5 above)
Comment (rossf7) [reply service/resource/s3bucketv1/resource.go]: No I'd missed removing it. Thanks its gone now.

**Pull request merged**